File: Matlab/Functions/read_Traces.m
function structure = read_Traces(path)

files=dir(fullfile(path,'*.json'));  % Reads all the json files in the directory.
no_files=length(files);              % Number of files in the directory.
structure(no_files)=struct();        % Structure containing the information of the traces.

for j = 1:no_files
    name=files(j).name;                 % Access the name of the i-th file.
    name_aux=split(name,'.');           % Split the name
    structure(j).name=name_aux{1};     

    fid = fopen(strcat(path,name));
    raw = fread(fid,inf); 
    str = char(raw'); 
    fclose(fid); 
    val = jsondecode(str);              %Structure containing the information of the file

    structure(j).id=val.id;  
    no_reports=length(val.reports);     %Number of reports of the trace
    structure(j).no_reps=no_reports;
    features=struct();
    
    microServices=strings(1,no_reports);
    operations=strings(1,no_reports);
    threads=zeros(1,no_reports);
    HRT=zeros(1,no_reports);
    eventID=strings(1,no_reports);
    parentsID=cell(1,no_reports);
    entries=strings(1,no_reports);
    

    for i = 1:no_reports
      events=val.reports(i);
      
      if isfield(events{1},'Operation')
           operations(i)=events{1}.Operation;
      end

      if isfield(events{1},'ParentEventID')
           parentsID{i}=events{1}.ParentEventID;
      else
           parentsID{i}='';
      end
      
      if isempty(events{1}.ProcessName)
          microServices(i)='Reference';
      else
          microServices(i)=events{1}.ProcessName;
      end
      
      entries(i)=events{1}.Agent;
      threads(i)=events{1}.ThreadID;
      eventID(i)=events{1}.EventID;
      HRT(i)=events{1}.HRT;
      
    end

    features.Operations=operations;
    features.Threads=threads;
    features.Services=microServices;
    features.eventID=eventID;
    features.parentsID=parentsID;
    features.HRT=HRT;
    features.Agents=entries;

    structure(j).feats=features;
    structure(j).initial_Time=min(HRT);
    structure(j).last_Time=max(HRT);

end
end

File: Python/read_traces.py
#Servicio social. Martín Guzmán Ruelas.

import os
import json
import pandas as pd

#replicating read_traces function
def read_traces(path):

    files = os.listdir(path) #a list with all the .json files
    structure1 = pd.DataFrame(columns = ['name', 'id', 'no_reps', 'initial_time', 'last_time']) #a dictionary containing the information of files
    structure2 = [] #a list that contains a dataframe for each file; this dataframe contains features information
    
    row = 0 #indicator
    
    for file in files:
        #we open the file in order to read its information
        with open(path + file) as current_file:
            current_data = json.load(current_file)[0]
        
        #creating the dataframe of features
        no_reports = len(current_data['reports'])
        
        #features
        operations = []
        threads = []
        eventID = []
        parentsID = []
        HRT = []
        agents = []
        microservices = []
        
        for i in range(no_reports):
            events = current_data['reports'][i]
            
            if 'Operation' in events:
                operations.append(events['Operation'])
            else:
                operations.append('')
            
            if 'ParentEventID' in events:
                parentsID.append(events['ParentEventID'])
            else:
                parentsID.append('')
            
            if events['ProcessName'] == '':
                microservices.append('Reference')
            else: 
                microservices.append(events['ProcessName'])
            
            agents.append(events['Agent'])
            threads.append(events['ThreadID'])
            eventID.append(events['EventID'])
            HRT.append(events['HRT'])
            
        features = {'Operations': operations,
                    'Threads': threads,
                    'Services': microservices,
                    'EventID': eventID,
                    'ParentID':parentsID,
                    'HRT': HRT,
                    'Agents':agents}
        
        structure2.append(pd.DataFrame(features))
            
        structure1.loc[row] = {'name'         : file,
                               'id'           : current_data['id'],
                               'no_reps'      : len(current_data['reports']),
                               'initial_time' : min(HRT),
                               'last_time'    : max(HRT)}
        
        row = row + 1
        
    return structure1, structure2

